experimental/adaptive-dialog/docs/recognizers-rules-steps-reference.md

```diff
@@ -0,0 +1,739 @@
+greetUserDialog.AddRule(new IntentRule("greetUser", 
+    steps: new List<IDialog>() {
+        new SendActivity("Hello")
+}));
+
+// Example that includees reference to property on bot state.
+var greetUserDialog = new AdaptiveDialog("greetUserDialog");
+greetUserDialog.AddRule(new IntentRule("greetUser", 
+    steps: new List<IDialog>() {
+        new TextInput("user.name", "What is your name?"),
+        new SendActivity("Hello, {user.name}")
+}));
+```
+See [here][3] to learn more about using language generation instead of hard coding actual response text in SendActivity.
+
+### SaveEntity
+Use this step to promte an entity (from a recognizer) into a different memory scope. By default entities from recognizer are available under the [turn scope][4] and the life time of all information under that scope is the end of that turn of conversation. 
+
+``` C#
+var greetUserDialog = new AdaptiveDialog("greetUserDialog");
+greetUserDialog.AddRule(new IntentRule("greetUser", 
+    steps: new List<IDialog>() {
+        // Save the userName entitiy from a recognizer.
+        new SaveEntity("user.name", "turn.entities.userName[0]"),
+        // Ask user for their name. All inputs by default will only initiate a prompt if the property does not exist.
+        new TextInput()
+        {
+            Prompt = new ActivityTemplate("What is your name?"),
+            Property = "user.name"
+        },
+        new SendActivity("Hello, {user.name}")
+}));
+```
+
+### EditArray
+Used to perform edit operations on an array property.
+
+``` C#
+var addToDoDialog = new AdaptiveDialog("addToDoDialog");
+addToDoDialog.AddRule(new IntentRule("addToDo", 
+    steps: new List<IDialog>() {
+        // Save the userName entitiy from a recognizer.
+        new SaveEntity("dialog.addTodo.title", "turn.entities.todoTitle[0]"),
+        new TextInput()
+        {
+            Prompt = new ActivityTemplate("What is the title of your todo?"),
+            Property = "dialog.addTodo.title"
+        },
+        // Add the current todo to the todo's list for this user.
+        new EditArray(EditArray.ArrayChangeType.Push, "user.todos", "dialog.addTodo.title"),
+        new SendActivity("Ok, I have added {dialog.addTodo.title} to your todos."),
+        new SendActivity("You now have {count(user.todos)} items in your todo.")
+}));
+```
+
+### InitProperty
+Initializes a property in memory. Can either initialize an array or object.
+
+``` C#
+new InitProperty()
+{
+    Property = "user.todos",
+    Type = "array" // this can either be "array" or "object"
+}
+```
+
+### SetProperty 
+Used to set a property's value in memory. The value can either be an explict string or an expression. See [here][5] to learn more about the common expressions language.
+
+``` C#
+new SetProperty() 
+{
+    Property = "user.firstName",
+    // If user name is Vishwac Kannan, this sets first name to 'Vishwac'
+    Value = new ExpressionEngine().Parse("split(user.name, ' ')[0]")
```

Comment:
similar to before, someone might ask why I need an expression vs. just writing code?